### ISSUE TITLE
Windows app fails when using sprite particles

### ISSUE THREAD

**Issue by IlyaFaer** [panda3d/panda3d#1106]
## Description
Hi, everyone!

I'm experiencing a program fail when trying to run a game built into win_amd64 with sprite particles. When running from Python source code everything is okay.

The problem is that I don't see any traceback/logs of the error - it just silently dies. Still, if there is something that I can try to provide more info, let me know, I'll try.

## Steps to Reproduce
- Generate a PTF file (with sprite renderer) with Panda3D particles GUI (attaching my file code lower).
- Use this particle as usual:
```python
snow = ParticleEffect()
snow.loadConfig("effects/snow.ptf")
snow.setPos(0, 0, 0.1)
snow.start(base.cam, render)  # noqa: F821
```
- Build the game into win_amd64 and run the resulting EXE file.

## Environment
 * Windows 10 x64
 * Panda3D version: 1.10.8 x64
 * Installation method: SDK
 * Python version: 3.7

**My PTF file**:
```
self.reset()
self.setPos(0.000, 0.000, 0.000)
self.setHpr(0.000, 0.000, 0.000)
self.setScale(1.000, 1.000, 1.000)
p0 = Particles.Particles('particles-1')
# Particles parameters
p0.setFactory("PointParticleFactory")
p0.setRenderer("SpriteParticleRenderer")
p0.setEmitter("RectangleEmitter")
p0.setPoolSize(100)
p0.setBirthRate(1.0000)
p0.setLitterSize(5)
p0.setLitterSpread(0)
p0.setSystemLifespan(0.0000)
p0.setLocalVelocityFlag(1)
p0.setSystemGrowsOlderFlag(0)
# Factory parameters
p0.factory.setLifespanBase(7.0000)
p0.factory.setLifespanSpread(1.0000)
p0.factory.setMassBase(1.0000)
p0.factory.setMassSpread(0.3000)
p0.factory.setTerminalVelocityBase(400.0000)
p0.factory.setTerminalVelocitySpread(0.0000)
# Point factory parameters
# Renderer parameters
p0.renderer.setAlphaMode(BaseParticleRenderer.PRALPHAOUT)
p0.renderer.setUserAlpha(1.00)
# Sprite parameters
p0.renderer.addTextureFromFile('effects/snowflake.png')
p0.renderer.setColor(Vec4(1.00, 1.00, 1.00, 1.00))
p0.renderer.setXScaleFlag(1)
p0.renderer.setYScaleFlag(1)
p0.renderer.setAnimAngleFlag(0)
p0.renderer.setInitialXScale(0.003)
p0.renderer.setFinalXScale(0.003)
p0.renderer.setInitialYScale(0.003)
p0.renderer.setFinalYScale(0.003)
p0.renderer.setNonanimatedTheta(0.0000)
p0.renderer.setAlphaBlendMethod(BaseParticleRenderer.PPBLENDLINEAR)
p0.renderer.setAlphaDisable(0)
# Emitter parameters
p0.emitter.setEmissionType(BaseParticleEmitter.ETEXPLICIT)
p0.emitter.setAmplitude(1.0000)
p0.emitter.setAmplitudeSpread(0.0000)
p0.emitter.setOffsetForce(Vec3(0.0000, 0.0000, -1.0000))
p0.emitter.setExplicitLaunchVector(Vec3(-0.6000, 0.0000, 0.0000))
p0.emitter.setRadiateOrigin(Point3(0.0000, 0.0000, 0.0000))
# Rectangle parameters
p0.emitter.setMinBound(Point2(-2.0000, -2.0000))
p0.emitter.setMaxBound(Point2(2.0000, 2.0000))
self.addParticles(p0)
f0 = ForceGroup.ForceGroup('noise')
# Force parameters
force0 = LinearNoiseForce(0.0500, 0)
force0.setVectorMasks(1, 1, 0)
force0.setActive(1)
f0.addForce(force0)
self.addForceGroup(f0)
```

**Comment by rdb** [comment]
> The problem is that I don't see any traceback/logs of the error - it just silently dies. Still, if there is something that I can try to provide more info, let me know, I'll try.

You need to set `log_filename` in your setup.py to get log output.

You have not shared your setup.py file—have you added `**/*.ptf` to your `include_patterns`?

**Comment by IlyaFaer** [comment]
@rdb, yeah, I can confirm that PTF files are okay in my build - I've tried the same PTF, but with Geom Renderer, it's working fine.

I also tried to use `log_filename`, but it gives me only this:
```
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
```
In fact in normal case I have such lines in my `main.py`:
```python
try:
    GameApp().run()
except Exception:
    logging.exception("Exception occured:")
```
It catches all the errors and logs them into the file I've set in `logging` configs. This current crash is the first one, which is not logged anywhere.

**Comment by rdb** [comment]
I would suggest that you just use the `log_filename` mechanism and disable your custom logging, I wonder if they are interfering.  If an exception is happening, you should see it in the log file specified with `log_filename`.

You can alternatively disable `log_filename` and build it via `console_apps` instead of `gui_apps`.  That will allow you to see the errors on the command prompt as usual.

**Comment by IlyaFaer** [comment]
@rdb,

> I wonder if they are interfering.

Hm-m, I didn't think about it. That could be so, but looks like not. I've tried to erase my custom logging and just use `log_filename` (both with gui and console apps) - the result is the same:
```
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
```

Putting my `setup.py` file here, maybe that'll give something:
```python
from setuptools import setup

setup(
    name="Forward Only",
    version="0.56",
    author="IlyaFaer",
    author_email="ilya.faer@mail.ru",
    description="Forward Only, the game",
    license="https://github.com/IlyaFaer/ForwardOnlyGame/blob/master/LICENSE.md",
    url="https://github.com/IlyaFaer/ForwardOnlyGame",
    options={
        "build_apps": {
            "platforms": ["win_amd64", "macosx_10_6_x86_64", "manylinux1_x86_64"],
            "include_patterns": [
                "**/*.bam",
                "**/*.jpg",
                "**/*.jpeg",
                "**/*.ogg",
                "**/*.png",
                "**/*.ptf",
                "**/*.ttf",
            ],
            "include_modules": {"*": ["__builtin__", "shelve", "dbm"]},
            "gui_apps": {"Forward Only": "main.py"},
            "plugins": ["pandagl", "p3fmod_audio", "p3openal_audio"],
        }
    },
)
```
Does sprite particles require some kind of a lib or something? Maybe it's missing in my build?
Still, I remember, when `p3fmod_audio` was missing in my build, I had an error log...

**Comment by rdb** [comment]
I'm unable to reproduce this on Linux using your given ptf file, but I will try on Windows later.

Unless you get a crash window ("This application has stopped working") I still think it must be an issue

You could also try explicitly adding `direct.particles.ParticleManagerGlobal` and `direct.showbase.PhysicsManagerGlobal` to your `include_modules`.

**Comment by IlyaFaer** [comment]
@rdb, yeal, I'll try it. Thanks!

**Comment by rdb** [comment]
I just built the Particle sample program without problem on Windows.  Please provide additional information or a test case.

**Comment by IlyaFaer** [comment]
@rdb, I've tried to build it once again today, and it's failing. What if I'll push a [branch in my repository](https://github.com/IlyaFaer/ForwardOnlyGame/tree/sprites_problem)!? If I'm building an app from this branch, it fails without any log.

Here is my setup.py:
```python
from setuptools import setup

setup(
    name="Forward Only",
    version="0.6",
    author="IlyaFaer",
    author_email="ilya.faer@mail.ru",
    description="Forward Only, the game",
    license="https://github.com/IlyaFaer/ForwardOnlyGame/blob/master/LICENSE.md",
    url="https://github.com/IlyaFaer/ForwardOnlyGame",
    options={
        "build_apps": {
            "platforms": [
                "win_amd64",
                "win32",
                "macosx_10_6_x86_64",
                "manylinux1_x86_64",
            ],
            "include_patterns": [
                "**/*.bam",
                "**/*.jpg",
                "**/*.jpeg",
                "**/*.ogg",
                "**/*.png",
                "**/*.ptf",
                "**/*.ttf",
            ],
            "include_modules": {"*": ["__builtin__", "shelve", "dbm"]},
            "gui_apps": {"Forward Only": "main.py"},
            "plugins": ["pandagl", "p3fmod_audio", "p3openal_audio"],
        }
    },
)
```

The PTF file with sprites is `effects/smoke_tail.ptf`. Here is [how it's used](https://github.com/IlyaFaer/ForwardOnlyGame/blob/45a98268554b85ae60dd7535003dff65f824ce60/world/objects.py#L228-L231).

**Comment by rdb** [comment]
@IlyaFaer you need to add a log_filename key to your build_apps configuration, otherwise you will not be able to see any logs.
Add this:
```python
            "log_filename": "$USER_APPDATA/Forward Only/output.log",
```
I also don't think you should be adding `__builtin__` to your `include_modules`.

**Comment by IlyaFaer** [comment]
@rdb, yeah, I've tried `log_filename` earlier, it didn't help. I still didn't try to explicitly add modules, like you adviced [here](https://github.com/panda3d/panda3d/issues/1106#issuecomment-798217860). And I think I'll try to build the app from Linux environment - who knows what'll happen.

**Comment by IlyaFaer** [comment]
I've tried to use explicit modules. And I also tried to build the app on Linux (which is very separate environment) - it still doesn't fails without log.

**Comment by rdb** [comment]
You *must always* build with log_filename when building a Windows app.

Please be explicit what you mean by "without log".  There is never a command-line output in a gui_apps application.  Are you saying that there is no file named output.log inside the Forward Only directory in AppData/Local?  Or does it exist but is it empty?  Or does it not even create the "Forward Only" directory inside AppData/Local?

What are the contents of your requirements.txt file?

**Comment by IlyaFaer** [comment]
@rdb, right now I've build the app with `console_apps` and `log_filename`.

The building session logs:
```
running bdist_apps
running build_apps
Building platforms: win_amd64
Gathering wheels for platform: win_amd64
Looking in indexes: https://pypi.org/simple, https://archive.panda3d.org/simple/opt, https://archive.panda3d.org/thirdparty
Collecting panda3d>=1.10.8
  Using cached https://buildbot.panda3d.org/downloads/90fb31ac219f7fb5e314686fed03cb069ef81554/opt/panda3d-1.10.8.post4%2Bopt-cp37-cp37m-win_amd64.whl
  Saved e:\git_reps\forwardonly\build\__whl_cache__\win_amd64_cp37m\panda3d-1.10.8.post4+opt-cp37-cp37m-win_amd64.whl
Successfully downloaded panda3d
Building runtime for platform: win_amd64
Unknown module: __builtin__
There are some missing modules: ['_dbm', '_posixsubprocess']
Copying game files for platform: win_amd64

Building zip for platform: win_amd64
```
The app is built, it's okay on a start (while in the game main menu). But when the particle effect starts, the game fails without any errors. The `log_filename` is there, but it contains only this:
```
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
```

Requirements of my app? It includes only single one line:
```panda3d>=1.10.8```

**Comment by rdb** [comment]
Hmm.  I tried building your game, and it ran.  I tried to create a new game, and it loaded, but then it seemed to hang, and I got "This application is not responding..."

Is there a way to launch the game in windowed instead of fullscreen mode?

**Comment by IlyaFaer** [comment]
@rdb, yeah, I also sometimes getting these `Unable to find texture` errors, but all the textures are loaded correctly.

To run the game windowed, you can comment this line:
https://github.com/IlyaFaer/ForwardOnlyGame/blob/7fa7ba5aaa056dd0161f9833591781eac2c489c7/main.py#L56

It'll be executed with default Panda3D window settings

**Comment by rdb** [comment]
I think the threading-model might have something to do with it.  I would suggest trying to comment it out and seeing if it still crashes/hangs.

**Comment by IlyaFaer** [comment]
@rdb, you're right indeed! If I'm commenting the threading-model line and building the app, it's working fine. Well, at least, now we know where is the root of the failure.

Seems like a common trouble, as it fails whereever and however I'm building it, meaning environments. I'll try not to use threading-model for now and see if FPS is on a good level, still, that would be good to have a fix. Anyway thanks already!

**Comment by IlyaFaer** [comment]
Hi!
Any ETAs on this issue?

I've found one more case in my game, when the same thing happens even without Sprite particles. I'm just loading the game, and it fails without any logs. The error is flaky - it fails 3/5 times. Turning off Cull/Draw makes it work fine, but I'm losing 2-3 FPS (not significant on my PC, but on others it can be worse I suppose).

The engine versions I've tried are 1.10.8, 1.10.9 and 1.10.10 - happens on all of them.
